src/java/com/puppetlabs/puppetserver/pool/JRubyPool.java

```diff
@@ -1,234 +1,355 @@
-    */
+     * Insert a poison pill into the pool.  It should only ever be used to
+     * insert a `PoisonPill` or `RetryPoisonPill` to the pool.
+     */
     @Override
-    public void insertPill(E e) throws InterruptedException {
-        liveQueue.putFirst(e);
+    public void insertPill(E e) {
+        final ReentrantLock lock = this.lock;
+        lock.lock();
+        try {
+            addFirst(e);
+        } finally {
+            lock.unlock();
+        }
     }
 
-    /**
-    * This method is analagous to <tt>clear</tt> in the
-    * <tt>LinkedBlockingDeque</tt> class.
-    *
-    * @throws InterruptedException
-    */
     @Override
     public void clear() {
-        liveQueue.clear();
+        final ReentrantLock lock = this.lock;
+        lock.lock();
+        try {
+            int queueSize = liveQueue.size();
+            for (int i=0; i<queueSize; i++) {
+                registeredElements.remove(liveQueue.removeFirst());
+            }
+        } finally {
+            lock.unlock();
+        }
     }
 
-    /**
-    * This method is analagous to <tt>remainingCapacity</tt> in the
-    * <tt>LinkedBlockingDeque</tt> class.
-    *
-    * @return the number of additional elements that the queue can ideally
-    *         accept without blocking.
-    */
     @Override
     public int remainingCapacity() {
-        return liveQueue.remainingCapacity();
+        int remainingCapacity;
+        final ReentrantLock lock = this.lock;
+        lock.lock();
+        try {
+            remainingCapacity = maxSize - liveQueue.size();
+        } finally {
+            lock.unlock();
+        }
+        return remainingCapacity;
     }
 
-    /**
-    * This method is analagous to <tt>size</tt> in the
-    * <tt>LinkedBlockingDeque</tt> class.
-    *
-    * @return the number of elements in the queue
-    */
     @Override
     public int size() {
-        return liveQueue.size();
+        int size;
+        final ReentrantLock lock = this.lock;
+        lock.lock();
+        try {
+            size = liveQueue.size();
+        } finally {
+            lock.unlock();
+        }
+        return size;
     }
 
-    /**
-    * Acquires a write lock from <tt>ReentrantReadWriteLock</tt>.
-    *
-    * Note that this implementation does not fulfill requirement (e) in the
-    * LockablePool javadocs for lock(); with this implementation, the thread
-    * holding the lock cannot perform a borrow while it's holding the lock.
-    */
     @Override
     public void lock() throws InterruptedException {
-        lock.writeLock().lock();
+        final ReentrantLock lock = this.lock;
+        lock.lock();
+        try {
+            final Thread currentThread = Thread.currentThread();
+            while (!isWriteLockHeldByCurrentThread(currentThread)) {
+                if (!isWriteLockHeld()) {
+                    writeLockThread = currentThread;
+                } else {
+                    notWriteLocked.await();
```

Comment:
@cprice404 and I talked about this a bit.  Contrary to the method signature, the implementation of this method on the master branch currently does not appear to lead to an `InterruptedException` being thrown.  To preserve the semantic behavior with the prior implementation, the line above and `allRegisteredInQueue.await()` call below could use `awaitUninterruptibly` instead.

Overall, though, I think it might be better, however, to just let this method be interruptible.  For example, a process shutdown might try to interrupt the thread.  Where these `await` calls could run for a pretty long time - e.g., while some particularly long code updates or catalog compilations are going on - it would seem better to just let the thread be interrupted so that the process can shut down sooner.  May need to rethink this later if we have more specific use cases for thread interruption across pool consumers.